### PR TITLE
Setup all the Entities #40013

### PR DESCRIPTION
**Pull request by shashankbaranwal**
- Create a versionable folder inside the repositories folder.
- Create a file VersionableSchema.ts and define generic schema fields inside it.
- Create a file IVersionableDocument.ts and define generic fields inside it.
- Extends above two in respective user schema and document.
- Create a file VersionableRepository that will include all the generic methods like create, update, genrateObjectId etc, It will be initialized with the model in the constructor and all the operations will be performed on the same model.
- Extend the same in UserRepository and use the generic methods for performing the operations.

### CODE REVIEW

File: src/repositories/user/UserRepository.ts [hunk 1]
```diff
@@ -1,26 +1,30 @@
...
+        console.log('User Data:', data);
         const id = UserRepository.generateObjectId();
         const model = new userModel({
             _id: id,
             ...data,
+            originalId: id,
         });
         return model.save();
     }
+
```
Comment (namanpar) [inline src/repositories/user/UserRepository.ts:27]: No need to create user for here instead use versionable create

File: src/libs/routes/validationHandler.ts [hunk 2]
```diff
@@ -1,54 +1,60 @@
...
+                error.push(a);
+                return;
+            }
+            if (inObject.isObject && (!(typeof value === 'object') || !(Object.entries(value).length))) {
+                a.key = keys;
+                a.location = inside;
+                a.errorMessage = `${keys} is invalid`;
+                error.push(a);
+                return;
+            }
```
Comment (namanpar) [inline src/libs/routes/validationHandler.ts:53]: do not object name as 'a, b, c' instead use temp or some well-defined vairable

**Review by namanpar — CHANGES_REQUESTED** [review]

File: src/repositories/versionable/VersionableRepository.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    }
+    public async userCreate(data: IUserModel): Promise<D> {
+        const id = VersionableRepository.generateObjectId();
+        const model = new this.model({
+            ...data,
+            _id: id,
+            originalId: id,
+        });
+        return await model.save();
+    }
```
[the remark this answers is not in the document]
Comment (shashankbaranwal) [reply src/repositories/versionable/VersionableRepository.ts:23]: Name Changed to create

File: src/repositories/versionable/VersionableRepository.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        if (!previous) {
+            return undefined;
+        }
+        await this.invalidate(data.originalId);
+        const newData = Object.assign(JSON.parse(JSON.stringify(previous)), data);
+        newData._id = VersionableRepository.generateObjectId();
+        delete newData.deletedAt;
+        const model = new this.model(newData);
+        return await model.save();
+    }
```
[the remark this answers is not in the document]
Comment (shashankbaranwal) [reply src/repositories/versionable/VersionableRepository.ts:59]: Name changed to create

File: src/repositories/user/UserSchema.ts [hunk 5]
```diff
@@ -1,4 +1,5 @@
 import * as mongoose from 'mongoose';
+import VersionableSchema from '../versionable/VersionableSchema';
 
 class UserSchema extends mongoose.Schema {
```
[the remark this answers is not in the document]
Comment (shashankbaranwal) [reply src/repositories/user/UserSchema.ts]: Extended the versionableSchema in UserSchema

File: src/repositories/user/UserRepository.ts (hunk 1 above)
Comment (shashankbaranwal) [reply src/repositories/user/UserRepository.ts]: changes done

File: src/libs/routes/validationHandler.ts (hunk 2 above)
Comment (shashankbaranwal) [reply src/libs/routes/validationHandler.ts]: Used temp variable

**Review by namanpar — APPROVED** [review]

**Pull request merged**